### PR TITLE
Add `width` and `height` options

### LINKED ISSUE #13: Finish #7
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/terminal-image/issues/13)

Would be nice to have https://github.com/sindresorhus/terminal-image/pull/7 finished, but the PR submitter is not responding.

Looking for someone to finish that PR and address the PR feedback. Preserve the existing commits and just add new ones in a new PR>

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/terminal-image/issues/13).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/terminal-image/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by Baidaly**
This PR adds`width` and `height` option from #7 and fixes #13

### CODE REVIEW

**Comment by Baidaly** [comment]
> @sindresorhus  Would you mind taking a look at the PR?

**Comment by sindresorhus** [comment]
There are some artifacts with this when the width gets small:

## master and your PR (in Terminal app)

Screenshot 2019-11-16 at 14 35 20Screenshot 2019-11-16 at 14 34 53

Notice the bottom row.

**Comment by sindresorhus** [comment]
While the output in master looks a bit more blurry, it feels like the output here looks less accurate:

Screenshot 2019-11-16 at 14 32 38Screenshot 2019-11-16 at 14 33 02

The first one is this PR and the second is `master`.

**Comment by sindresorhus** [comment]
Sorry for the very late review. I just have had way too many PRs to review...

**Comment by sindresorhus** [comment]
@Baidaly Still interested in finishing this? 🙏

**Comment by Baidaly** [comment]
@sindresorhus  Sorry missed your November comment. Yes, let me have a look at how to fix that. However, I will be able to work on it only next weekend. Hope it is OK.

**Comment by sindresorhus** [comment]
Great!

> However, I will be able to work on it only next weekend. Hope it is OK.

Not a problem at all :)

**Comment by Baidaly** [comment]
@sindresorhus Updated the code. The only thing I am not sure about is `preserveAspectRatio` behavior. I implemented as I understood based on your conversation in #7 but I might got it wrong.

**Comment by sindresorhus** [comment]
With:

```js
(async () => {
	console.log(await terminalImage.file('fixture.jpg', {width: 60, preserveAspectRatio: false}));
})();
```

I get:

Screenshot 2020-03-09 at 12 42 17

Which is not 60 column in width. So something about the calculation is wrong when `preserveAspectRatio: false`.

**Comment by sindresorhus** [comment]
You also need to add support for `width`/`height` accepting a percentage of the viewport, per https://github.com/sindresorhus/terminal-image/pull/7#issuecomment-389589497. And percentage should be recommended in the docs.

File: index.d.ts [hunk 1]
```diff
@@ -24,7 +24,11 @@ declare const terminalImage: {
 	Display images in the terminal.
 
 	@param filePath - File path to the image.
-	@returns The ansi escape codes to display the image.
+	@param options - Image rendering options.
+	@param options.width - Custom image width.
+	@param options.height - Custom image height.
```
Comment (sindresorhus) [inline index.d.ts]: You need to document what the default is for width/height. In the readme too.

File: readme.md [hunk 2]
```diff
@@ -26,6 +26,26 @@ const terminalImage = require('terminal-image');
 })();
 ```
 
+Optionally, you can specify the `height` and/or `width` to scale the image.
```
Comment (sindresorhus) [inline readme.md]: The new options needs to be properly documented in the `API` section. The readme and TypeScript file should be kept in sync.

File: readme.md [hunk 3]
```diff
@@ -26,6 +26,25 @@ const terminalImage = require('terminal-image');
...
+
+```js
+const terminalImage = require('terminal-image');
+
+(async () => {
+	console.log(await terminalImage.file('unicorn.jpg', {width: 1024, height: 1024}));
```
Comment (sindresorhus) [inline readme.md]: These example sizes are unrealistically large. The width/height are columns/rows, so use a smaller number.

File: index.d.ts (hunk 1 above)
Comment (sindresorhus) [reply index.d.ts]: Also note that width/height are columns/rows.

**Comment by Baidaly** [comment]
@sindresorhus Thank you for the reviewing the PR. Will work on it this week.

**Comment by snobu** [comment]
@Baidaly, @sindresorhus Willing to help here (full disclosure, i need this feature for https://github.com/snobu/destreamer). What can i do, just send a PR to update the README with the new height/width API? Anything else?

File: index.d.ts [hunk 4]
```diff
@@ -34,8 +38,15 @@ declare const terminalImage: {
 		console.log(await terminalImage.file('unicorn.jpg'));
 	})();
 	```
-	*/
-	file(filePath: string): Promise<string>;
+	 */
```
Comment (sindresorhus) [inline index.d.ts:41]:
```suggestion
	*/
```

File: index.js [hunk 5]
```diff
@@ -8,15 +8,56 @@ const termImg = require('term-img');
 const PIXEL = '\u2584';
 const readFile = util.promisify(fs.readFile);
 
-async function render(buffer) {
-	const image = await Jimp.read(buffer);
-	const columns = process.stdout.columns || 80;
-	const rows = process.stdout.rows || 24;
+function scale(width, height, originalWidth, originalHeight) {
+	const originalRatio = originalWidth / originalHeight;
+	const f = (width / height > originalRatio ? height / originalHeight : width / originalWidth);
```
Comment (sindresorhus) [inline index.js:13]: Use descriptive variable names.

File: readme.md [hunk 6]
```diff
@@ -26,6 +26,25 @@ const terminalImage = require('terminal-image');
...
+
+By default, aspect ratio is always maintained. If you don't want to maintain aspect ratio, set `preserveAspectRatio` to false. However, your image will be scaled to fit the size of the terminal.
+
+```js
+const terminalImage = require('terminal-image');
+
+(async () => {
+	console.log(await terminalImage.file('unicorn.jpg', {width: 1024, preserveAspectRatio: false}));
+})();
+```
```
Comment (sindresorhus) [inline readme.md:47]: This should be in index.d.ts too. The readme and it should be in sync.

**Comment by Baidaly** [comment]
@sindresorhus @snobu Updated the PR. Please have a look. The only thing I am not sure is if I documented the changes properly.

File: readme.md [hunk 7]
```diff
@@ -26,13 +26,24 @@ const terminalImage = require('terminal-image');
 })();
 ```
 
-Optionally, you can specify the `height` and/or `width` to scale the image.
+Optionally, you can specify the `height` and/or `width` to scale the image. That can be either the percentage of the terminal window or number of rows and/or columns. Please note that the image will always be scaled to fit the size of the terminal. If width and height are not defined, by default the image will take the width and height of the terminal.
 
+It is recommended to use the percentage option.
```
Comment (sindresorhus) [inline readme.md:31]: The readme and index.d.ts should be in sync.

File: index.js [hunk 8]
```diff
@@ -10,12 +10,27 @@ const readFile = util.promisify(fs.readFile);
...
-	height = f * originalHeight;
+	const factor = (width / height > originalRatio ? height / originalHeight : width / originalWidth);
+	width = factor * originalWidth;
+	height = factor * originalHeight;
 	return {width, height};
 }
 
+function checkAndGetDimensionValue(value, percentageBase) {
+	if (typeof value === 'string' && value.endsWith('%')) {
+		const percentageValue = parseFloat(value);
```
Comment (sindresorhus) [inline index.js:21]:
```suggestion
		const percentageValue = Number.parseFloat(value);
```

File: index.js [hunk 9]
```diff
@@ -10,12 +10,27 @@ const readFile = util.promisify(fs.readFile);
...
+		if (!isNaN(percentageValue) && percentageValue > 0 && percentageValue <= 100) {
```
Comment (sindresorhus) [inline index.js:22]:
```suggestion
		if (!Number.isNaN(percentageValue) && percentageValue > 0 && percentageValue <= 100) {
```

File: index.d.ts [hunk 10]
```diff
@@ -21,21 +21,40 @@ declare const terminalImage: {
...
-	Display images in the terminal.
+	Display images in the terminal. Please note that the image will always be scaled to fit the size of the terminal.
+	Optionally, you can specify the height and/or width to scale the image.
+	That can be either the percentage of the terminal window or number of rows and/or columns.
+	Please note that the image will always be scaled to fit the size of the terminal.
+	If width and height are not defined, by default the image will take the width and height of the terminal.
+	It is recommended to use the percentage option.
+	That can be either the percentage of the terminal window or number of rows and/or columns.
+	By default, aspect ratio is always maintained. If you don't want to maintain aspect ratio, set `preserveAspectRatio` to false.
+	However, your image will be scaled to fit the size of the terminal.
```
Comment (sindresorhus) [inline index.d.ts:32]: This is missing the examples from the readme.

Comment (Baidaly) [reply index.d.ts]: Updated the example section in the file

**Comment by Baidaly** [comment]
@sindresorhus Updated documentation

**Comment by sindresorhus** [comment]
The percentages are not working as expected. I expected `{width: '100%'}` to force it to take up 100% of the terminal width, even if it doesn't fit in the height of the terminal. Same with `{height: '100%'}`. It should take up 100% of the height (while preserving aspect ratio) of the terminal. Right now, specifying `{width: '100%'}` only makes the image as wide as it can fit in the terminal height.

Screenshot 2020-05-06 at 14 10 57

**Comment by sindresorhus** [comment]
The current behavior of fitting, should only apply if both width and height are specified with percentage (or not specified at all as it's the default values).

**Comment by Baidaly** [comment]
> The percentages are not working as expected. I expected `{width: '100%'}` to force it to take up 100% of the terminal width, even if it doesn't fit in the height of the terminal. Same with `{height: '100%'}`. It should take up 100% of the height (while preserving aspect ratio) of the terminal. Right now, specifying `{width: '100%'}` only makes the image as wide as it can fit in the terminal height.
> 
> Screenshot 2020-05-06 at 14 10 57

Changed it. Please have a look again.

**Comment by sindresorhus** [comment]
Looks good now. Thanks for persevering through this 🙌

**Comment by snobu** [comment]
Absolutely fabulous, thank you for introducing this! Is it up on npm yet?

**Comment by sindresorhus** [comment]
My laptop battery ran out while publishing... I’ll publish it in an hour or two.

**Comment by sindresorhus** [comment]
Done: https://github.com/sindresorhus/terminal-image/releases/tag/v1.0.0

**Pull request merged**